File: Models/popup.js
// File: robjam1990/Psychosis/Models/popup.js
// Import the Settings Controller module
import * as SettingsController from './settings-controller.js';

/**
 * Handle change events on the options form.
 * @param {SettingsController.Settings} settings
 */
function handleOptionsFormChange(settings) {
    settings = readSettingsFromDomAndPersist();
    configureVisibleSettings(settings);
}

/**
 * Configure visibility of settings based on backend.
 * @param {SettingsController.Settings} settings
 */
function configureVisibleSettings(settings) {
    const optionsCategoriesList = find('.options__categories');
    optionsCategoriesList.parentElement?.classList.toggle('hidden', settings.backend === 'psi');
}

/**
 * Read settings from the DOM and save them.
 * @return {SettingsController.Settings}
 */
function readSettingsFromDomAndPersist() {
    const backendInput = find('.options__backend input:checked');
    const backend = backendInput ? backendInput.value : 'defaultBackend';

    // Construct settings object with input validation
    const settings = {
        backend: backend,
        // Add validation for other settings here
    };

    // Save settings
    SettingsController.saveSettings(settings);

    return settings;
}

/**
 * Find an element in the DOM with a given query.
 * Throws an error if not found.
 * @template {string} T
 * @param {T} query
 * @param {ParentNode=} context
 * @return {HTMLElement}
 */
function find(query, context = document) {
    const result = context.querySelector(query);
    if (!result) {
        throw new Error(`Element with query "${query}" not found`);
    }
    return result;
}

/**
 * Initialize the popup's state and UI elements.
 */
async function initPopup() {
    try {
        const siteUrl = await getSiteUrl();
        const settings = await SettingsController.loadSettings();

        // Show browser-specific elements
        const browserBrandEl = find('.browser-brand--chrome'); // Assuming chrome for simplicity
        browserBrandEl.classList.remove('hidden');

        // Add event listener for generating report
        const generateReportButton = find('button.button--generate');
        generateReportButton.addEventListener('click', () => {
            onGenerateReportButtonClick(settings.backend, siteUrl.href, settings);
        });

        // Add event listener for options form change
        const optionsFormEl = find('.options__form');
        optionsFormEl.addEventListener('change', () => {
            handleOptionsFormChange(settings);
        });
    } catch (error) {
        handleInitializationError(error);
    }
}

/**
 * Get the site URL.
 * @return {Promise<URL>}
 */
function getSiteUrl() {
    return new Promise((resolve, reject) => {
        chrome.tabs.query({ active: true, lastFocusedWindow: true }, function (tabs) {
            if (tabs.length === 0 || !tabs[0].url) {
                reject(new Error('No active tab found'));
                return;
            }

            const url = new URL(tabs[0].url);
            if (url.hostname === 'localhost') {
                reject(new Error('Use DevTools to audit pages on localhost'));
            } else if (/^(chrome|about)/.test(url.protocol)) {
                reject(new Error(`Cannot audit ${url.protocol}// pages`));
            } else {
                resolve(url);
            }
        });
    });
}

/**
 * Handle initialization errors.
 * @param {Error} error
 */
function handleInitializationError(error) {
    const generateReportButton = find('button.button--generate');
    const psiDisclaimerEl = find('.psi-disclaimer');
    const errorMessageEl = find('.errormsg');

    generateReportButton.disabled = true;
    psiDisclaimerEl.remove();
    errorMessageEl.textContent = error.message;
}

// Initialize the popup
initPopup();

File: Models/popup.py
# File: robjam1990/Psychosis/Models/popup.py
# Import the Settings Controller module
import settings_controller

def handle_options_form_change(settings):
    settings = read_settings_from_dom_and_persist()
    configure_visible_settings(settings)

def configure_visible_settings(settings):
    options_categories_list = find('.options__categories')
    if options_categories_list.parentElement:
        options_categories_list.parentElement.classList.toggle('hidden', settings.backend == 'psi')

def read_settings_from_dom_and_persist():
    backend_input = find('.options__backend input:checked')
    backend = backend_input.value if backend_input else 'defaultBackend'

    # Construct settings object with input validation
    settings = {
        'backend': backend,
        # Add validation for other settings here
    }

    # Save settings
    settings_controller.save_settings(settings)

    return settings

def find(query, context=document):
    result = context.querySelector(query)
    if not result:
        raise ValueError(f'Element with query "{query}" not found')
    return result

async def init_popup():
    try:
        site_url = await get_site_url()
        settings = await settings_controller.load_settings()

        # Show browser-specific elements
        browser_brand_el = find('.browser-brand--chrome') # Assuming chrome for simplicity
        browser_brand_el.classList.remove('hidden')

        # Add event listener for generating report
        generate_report_button = find('button.button--generate')
        generate_report_button.addEventListener('click', lambda: on_generate_report_button_click(settings['backend'], site_url.href, settings))

        # Add event listener for options form change
        options_form_el = find('.options__form')
        options_form_el.addEventListener('change', lambda: handle_options_form_change(settings))
    except Exception as error:
        handle_initialization_error(error)

async def get_site_url():
    return new Promise((resolve, reject) => {
        chrome.tabs.query({ active: True, lastFocusedWindow: True }, lambda tabs: {
            if len(tabs) == 0 or not tabs[0].url:
                reject(ValueError('No active tab found'))
                return

            url = URL(tabs[0].url)
            if url.hostname == 'localhost':
                reject(ValueError('Use DevTools to audit pages on localhost'))
            elif re.match('^(chrome|about)', url.protocol):
                reject(ValueError(f'Cannot audit {url.protocol}// pages'))
            else:
                resolve(url)
        })

def handle_initialization_error(error):
    generate_report_button = find('button.button--generate')
    psi_disclaimer_el = find('.psi-disclaimer')
    error_message_el = find('.errormsg')

    generate_report_button.disabled = True
    psi_disclaimer_el.remove()
    error_message_el.textContent = error.message

# Initialize the popup
init_popup()
